### PR TITLE
Introduce lite-mode

### PR DESCRIPTION
**Pull request by lordshashank**
## Usage related changes

Would implement lite node that can be used with "--lite-node" tag which would skip blockhash calculation by replacing it with block_number hashes.

## Development related changes

- Resolves https://github.com/0xSpaceShard/starknet-devnet-rs/issues/343
- Added a `lite_mode` option in the configuration.
- Modified the `generate_new_block` function to skip block hash calculation when `lite_mode` is enabled.
- Added relevant docs for lite-mode.

## Benchmark Results

I performed benchmark testing using [mint_benchmark[(./scripts/benchmark/mint to compare the performance of the application in `lite_mode` and regular mode. Here are the results:

| Mode        | Command | Number of Mints | Time (ms) |
|-------------|---------|-----------------|-----------|
| Lite Mode   | cargo run -- --lite-mode | 200 |   11.172     |
| Regular Mode| cargo run | 200 |    13.447   |
| Lite Mode (Release) | cargo run --release -- --lite-mode | 1000 |   14.703   |
| Regular Mode (Release) | cargo run --release | 1000 |    15.058   |

As expected, enabling `lite_mode` improves the performance of the application.

## Checklist:

- [ ] Checked the relevant parts of [development docs](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#development)
- [X] Applied formatting - `./scripts/format.sh`
- [X] No linter errors - `./scripts/clippy_check.sh`
- [X] No unused dependencies - `./scripts/check_unused_deps.sh`
- [X] Performed code self-review
- [X] Rebased to the latest commit of the target branch (or merged it into my branch)
- [X] Updated the docs if needed, including the [TODO section](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#todo-to-reach-feature-parity-with-the-pythonic-devnet)
- [X] Linked the [issues](https://github.com/0xSpaceShard/starknet-devnet-rs/issues) which this PR resolves
- [ ] Updated the tests if needed; all passing ([execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/?tab=readme-ov-file#test-execution))

### CODE REVIEW

**Comment by lordshashank** [comment]
@ivpavici I have implemented lite mode by skipping blockhash calculation [here](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/444/files#diff-c3e1711bf66de593f5875f51b8500529fc04c7a785ffb1c499b0407b69f6cc9aR279) and replacing it with hash of block_number. 

Could you please see if the implementation is correct, I am not sure of few things like would calculating [block_number](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/444/files#diff-c3e1711bf66de593f5875f51b8500529fc04c7a785ffb1c499b0407b69f6cc9aR275) before initializing header cause problem?, validity of initializing hash with ```from_prefixed_hex_string(block_number_hex)``` to reduce computation than ```generate_hex()```, or if I have went completely wrong way here. 

Also let me know if hash calculation has to be skipped somewhere else as well, thanks.

**Review by FabijanC — COMMENTED** [review]
Have you done any benchmarking? What is the time impact of skipping block hash calculation?

**Comment by lordshashank** [comment]
Not yet, I just implemented the functionality as per devnet-py, if its required let me know how I can do that?

**Comment by FabijanC** [comment]
> Not yet, I just implemented the functionality as per devnet-py, if its required let me know how I can do that?

Think of something, be creative :D

But one test could look like this, let's call the script `mint_benchmark.sh`:
```bash
#!/bin/bash

set -eu

N=$1

for _ in $(seq 1 $N); do
    curl -w "\n" -sSf localhost:5050/mint --json '{
        "amount": 1,
        "address": "0x1"
    }'
done
```

And you could run:
```
$ chmod +x mint_benchmark.sh
$ time ./mint_benchmark.sh 1000
```

Or replace 1000 with any number you wish to test. Of course, this assumes that somewhere on your computer you have Devnet running at port 5050. I'd recommend testing both `cargo run` and `cargo run --release`, both in lite and regular mode.

**Comment by ivpavici** [comment]
yeah, please @lordshashank we would be very grateful for this test and report afterwards :)

**Comment by lordshashank** [comment]
@FabijanC @ivpavici I have added benchmark report and test file as well, let me know if something else is needed.

**Comment by ivpavici** [comment]
thanks!
Seems like the speedup is negligible in  `--release` mode... I will let @FabijanC decide about the merging. But either way this was helpful for us so you will definitely get a reward regardless if we merge or not!

File: README.md [hunk 1]
```diff
@@ -366,6 +367,14 @@ Response:
 }
 ```
 
+## Lite mode
```
Comment (FabijanC) [inline README.md:370]:
```suggestion
## Lite Mode
```

File: README.md [hunk 2]
```diff
@@ -366,6 +367,14 @@ Response:
...
+
+Runs the Devnet in a minimal lite mode by just skipping the block hash calculation. This is useful for testing purposes when the block hash is not needed.
```
Comment (FabijanC) [inline README.md:372]:
```suggestion
Runs Devnet in a minimal lite mode by just skipping the block hash calculation. This is useful for testing purposes when the block hash is not needed.
```

File: crates/starknet-devnet/src/cli.rs [hunk 3]
```diff
@@ -123,6 +123,12 @@ pub(crate) struct Args {
     #[arg(requires = "dump_path")]
     dump_on: Option<DumpOn>,
 
+    #[arg(long = "lite-mode")]
+    #[arg(value_name = "LITE_MODE")]
```
Comment (FabijanC) [inline crates/starknet-devnet/src/cli.rs:127]: This has no effect for boolean flags, does it?

File: scripts/benchmark/mint_benchmark.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -eu
+
+N=$1
+ADDRESS=$2
```
Comment (FabijanC) [inline scripts/benchmark/mint_benchmark.sh:6]: Address needn't be a parameter

File: scripts/benchmark/mint_benchmark.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+for _ in $(seq 1 $N); do
+    curl -w "\n" -sSf -H "Content-Type: application/json" -d '{
+        "amount": 1,
+        "address": "'"$ADDRESS"'"
+    }' localhost:5050/mint
+done
```
Comment (FabijanC) [inline scripts/benchmark/mint_benchmark.sh:13]: Add a newline

File: crates/starknet-devnet-core/src/starknet/mod.rs [hunk 6]
```diff
@@ -288,18 +288,26 @@ impl Starknet {
...
         // set new block header
-        new_block.set_block_hash(new_block.generate_hash()?);
+        if self.config.lite_mode {
+            new_block.set_block_hash(BlockHash::from_prefixed_hex_str(&format!(
+                "0x{:x}",
+                new_block_number.0
+            ))?);
+        } else {
+            new_block.set_block_hash(new_block.generate_hash()?);
+        }
```
Comment (FabijanC) [inline crates/starknet-devnet-core/src/starknet/mod.rs:302]: Can be shortened to `new_block.set_block_hash(if self.config.lite_mode { ... } else { ... })`

File: crates/starknet-devnet-core/src/starknet/mod.rs [hunk 7]
```diff
@@ -288,18 +288,26 @@ impl Starknet {
...
         let mut new_block = self.pending_block().clone();
 
+        let new_block_number =
+            BlockNumber(new_block.block_number().0 - self.blocks.aborted_blocks.len() as u64);
+
         // set new block header
-        new_block.set_block_hash(new_block.generate_hash()?);
+        if self.config.lite_mode {
+            new_block.set_block_hash(BlockHash::from_prefixed_hex_str(&format!(
+                "0x{:x}",
```
Comment (FabijanC) [inline crates/starknet-devnet-core/src/starknet/mod.rs:297]: Can be `"{:#x}"`

**Review by FabijanC — CHANGES_REQUESTED** [review]
Thanks for the table report in the PR description! How did you accumulate the times? I mean - are you reporting what you got on the first run? Or did you run it multiple times?

Also, I'm having doubts about DX and UX and if `--lite-mode` is the best approach. Perhaps we should have multiple CLI options like `--block-hash <REAL|DUMMY>`. We should consider what else can be skipped (other than block hash generation).

**Review by FabijanC — APPROVED** [review]
Thanks!

**Comment by FabijanC** [comment]
@lordshashank Please, in the future, execute the tests to ensure the CI workflow passes. I will fix the failure coming from the merge of this PR.

**Comment by lordshashank** [comment]
Really really sorry, I was thinking to do that after all the issues in pr were resolved, just forgot it at the end. Would surely keep in mind from next time.

**Pull request merged**